target/linux/ramips/dts/RE650.dts

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "tplink,re650-v1", "mediatek,mt7621-soc";
+	model = "TP-Link RE650 v1";
+
+	aliases {
+		led-boot = &power;
+		led-failsafe = &power;
+		led-running = &power;
+		led-upgrade = &power;
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x8000000>;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		power: power {
+			label = "re650-v1:blue:power";
+			gpios = <&gpio0 22 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan2g {
+			label = "re650-v1:blue:wlan2g";
+			gpios = <&gpio0 23 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan5g {
+			label = "re650-v1:blue:wlan5g";
+			gpios = <&gpio0 24 GPIO_ACTIVE_LOW>;
+		};
+
+		wps_red {
+			label = "re650-v1:red:wps";
+			gpios = <&gpio0 25 GPIO_ACTIVE_HIGH>;
+		};
+
+		wps_blue {
+			label = "re650-v1:blue:wps";
+			gpios = <&gpio0 26 GPIO_ACTIVE_HIGH>;
+		};
+
+		eth_act {
+			label = "re650-v1:green:eth_act";
+			gpios = <&gpio0 28 GPIO_ACTIVE_LOW>;
+		};
+
+		eth_link {
+			label = "re650-v1:green:eth_link";
+			gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		wps {
+			label = "wps";
+			gpios = <&gpio0 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+
+		led {
+			label = "led";
+			gpios = <&gpio0 30 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_LIGHTS_TOGGLE>;
+		};
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 31 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		power {
+			label = "power";
+			gpios = <&gpio1 0 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_POWER>;
+		};
+
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	w25q64@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x20000>;
+				read-only;
+			};
+
+			partition@20000 {
+				compatible = "tplink,firmware";
+				label = "firmware";
+				reg = <0x20000 0xde0000>;
+			};
+
+			config: partition@e00000 {
+				label = "config";
+				reg = <0xe00000 0x50000>;
+				read-only;
+			};
+
```

Comment:
I'm not familiar with these devices: Is 0xe50000 to 0xff0000 really "empty"?
(tplink-safeloader and partitions here are matching)

Reply:
Yes, it's empty. There's nothing on the flash in that range (0xe50000-0xff0000).

TP-Link partition table:
Creating 5 MTD partitions on "raspi":                                                                                                                                                                               
0x000000000000-0x000001000000 : "ALL"                                                                                                                                                                               
0x000000000000-0x000000020000 : "fs-uboot"                                                                                                                                                                          
0x000000020000-0x000000310000 : "os-image"                                                                                                                                                                          
0x000000310000-0x000000e00000 : "file-system"                                                                                                                                                                       
0x000000ff0000-0x000001000000 : "radio"

Reply:
Just have seen it elsewhere and it seems to be a good idea:
Add a comment to the DTS stating this area is empty, e.g.
`\* range 0x5e0000 to 0xff0000 is empty in vendor firmware, so we do not use it either *\`